### DISCUSSION TITLE
Support for text folding via hs-minor-mode

### DISCUSSION THREAD

**Discussion by mplanchard** [joaotavora/eglot#1431] [Ideas]
It would be nice if eglot plugged into `hs-minor-mode` in order to provide code folding functionality, which I think would use the [textDocument/foldingRange](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_foldingRange) provider in language servers that support it.

**Comment by joaotavora** [comment]
That _would_ be nice, I agree.  

Unfortunately, last I look, `hs-minor-mode` didn't have a programmable API for Eglot to plug in the LSP information.  This is not uncommon in Emacs land.  Practically before starting Eglot, I redesigned Flymake to allow that.  You need to bite the bullet and write a a `hs2-minor-mode` that copies that UI things of `hs-minor-mode`  (which are reasonably good, like isearch integration, etc) but supports the LSP info (or other types of folding info).

Then do the same for all the other stuff: class diagrams, code lens, etc :-D

**Reply by mplanchard** [reply]
Well that's sad news. Thank you for the response though!

One thing is that `treesit-fold` works quite well for treesitter-enabled modes. Unfortunately for my main language (rust) treesitter mode currently breaks rustic's run-current-test functionality (https://github.com/brotzeit/rustic/issues/571), so I have to hack it in with:

```elisp
(use-package treesit-fold
  :ensure (:type git :host github :repo "emacs-tree-sitter/treesit-fold")
  :hook
  (rustic-mode . ((treesit-parser-create 'rust))
  :config
  (global-treesit-fold-mode))
```

It's a pretty nice package, but I like eglot's general philosophy of using emacs' builtin stuff wherever possible. It's possible that eventually treesitter will be widespread enough that LSP folding won't even be necessary I suppose

**Reply by joaotavora** [reply]
There's built in tree-sitter integration in core which is completely parallel to this one, including this util. 

This is a completely different different situation, albeit analogously sad.

Eglot can have small amounts of integration code for some optional non Emacs-core things when it makes sense, is simple and provides value (examples: yasnippet, company, breadcrumb). 

So if "they" write a hypothetical "foldyfoldy.el" mode that is robust and generic enough I could accept a PR for some kind of LSP integration.
